Output space exploration

quaquel opened the pull request:
adds output space exploration to the workbench. 

- [x] unit tests and documentation are in place.
- [x] I might still want to update the example notebook
- [x] unit tests adds pytest-mock as a dependency in ci, and thus also for #166 

Closes #154.

EwoutH commented:
Looks awesome! I will review it from a code-quality and user perspective tomorrow.

EwoutH commented:
I started on a review, in general, it looks quite good!

Reviewing the docs on the other hand is quite a challenge without them begin built. Could you check if it's easy to enable doc building (see #171) before I review them?

quaquel commented:
I have enabled the building of docs for pull requests. I guess we need one more commit here to trigger it.

EwoutH commented:
If you can find the time, [cleaning up you commits](https://dev.to/aman_singh/how-to-cleanup-commit-history-using-git-reset-46p3) would be a great way to also trigger the CI! :)

Otherwise I will do some ci-trigger-magic when I review it (probably tomorrow).

EwoutH commented:
Updated the CI and pyproject.toml with pytest-mock.

Read the Docs now runs, and generates the docs for this PR. However, for some reasone the outputspace_exploration module doesn't appear in it: https://emaworkbench--170.org.readthedocs.build/en/170/api_index.html#exploratory-modeling-framework

quaquel commented:
Ok, I'll have a look later today. Probably missing links in some index files.

EwoutH commented:
Perfect, take your time. I started on the review of the Jupyter example notebook, I hope to finish it later today.

quaquel commented:
the docs now are displaying properly. I guess this is ready for a last cleanup of the commit history before being merged, unless you find something in the notebook

EwoutH commented:
I just reviewed the notebook, and it's structured quite well! The main thing I'm missing is a bit of explanation and interpretation, from which the viewer can gain a better understanding of this method. It would add a lot of value if this example would be readable on itself, as a primer to OSE.

Reviewing Jupyter notebooks in GitHub turned out to be quite cumbersome, so I searched a bit and found an app, ReviewNB, which made this easy. I left some copies on the notebok, you can view and reply here (you probably have to login with GitHub): https://app.reviewnb.com/EwoutH/EMAworkbench/blob/outputspace_review/ema_workbench%2Fexamples%2Foutputspace_exploration_lakeproblem.ipynb

quaquel commented:
That review app looks nice and the suggestions will indeed make the notebook a lot better. I'll try to make the suggested changes over the coming days.

quaquel commented:
I have updated the notebook broadly in line with your suggestions. Note that I don't intend for these notebooks to be a tutorial on MOEAs so I have kept the additional text short but with links to relevant literature in case you want to know more.

I also added a small check to ensure the number of items in `grid_spec` matches the number of objectives (analogous to the problem with epsilon values in e-NSGA2). This in turn necessitated a small update to the test code.

quaquel commented:
I cleaned the commit history again, so if you agree with the notebook update, this is ready to be merged.

EwoutH reviewed (changes requested):
Code quality and structure looks good, I didn't encounter strange things.

Docs are also coming along nicely, I added a few comments on things I found unclear or ambiguous.

If you take a look at those, could you let me merge afterwards? I will squash it into a nice tidy commit.

quaquel commented on ema_workbench/em_framework/optimization.py:
```diff
@@ -266,7 +266,9 @@ def to_dataframe(optimizer, dvnames, outcome_names):
     """
 
     solutions = []
-    for solution in platypus.unique(platypus.nondominated(optimizer.result)):
```
First, it is redundant. Optimizer.results already returns the non dominated solutions. Second, it breaks with output space exploration, where we do not want to include directional information.

quaquel commented on ema_workbench/em_framework/evaluators.py:
```diff
@@ -716,6 +719,7 @@ def optimize(
                         nfe between convergence check
     logging_freq : int
                    number of generations between logging of progress
+    variator : platypus GAOperator instance, optional
```
Not sure whether I want to go there. By defaulting to None, we trigger the default values within platypus-opt. The default in platypus-opt is Simulated Binary Crossover with Polynomial Mutation. Any custom platypus-opt GAOperator instance is possible, but what that means is covered by platypus-opt.

EwoutH replied:
In that case I think I don't understand the _optional_. What happens if you leave `variator` unspecified?

quaquel replied:
Then it takes its default value, which is None, as with any other keyword argument.

EwoutH replied:
So should we mention that it uses the platypus-opt default, which is Simulated Binary Crossover with Polynomial Mutation?

quaquel replied:
I added a bit to the documentation

quaquel commented on ema_workbench/em_framework/outputspace_exploration.py:
```diff
@@ -0,0 +1,409 @@
+
+
+class AutoAdaptiveOutputSpaceExploration(AdaptiveTimeContinuation):
+    """A combination of autoadaptive operator selection with OutputSpaceExploration.
```
problem and grid spec are necessary. The rest has meaningful defaults

quaquel commented on ema_workbench/em_framework/outputspace_exploration.py:
```diff
@@ -0,0 +1,409 @@
+
+The algorithm can be used in combination with the optimization functionality of the workbench.
+Just pass an OutputSpaceExploration instance as algorithm to optimize.
+
```
If I know how, then yes. This probably overlaps with adding an overview page with links to all examples

quaquel commented on ema_workbench/em_framework/outputspace_exploration.py:
```diff
@@ -0,0 +1,409 @@
+)
+
+
+__all__ = ["OutputSpaceExploration", "AutoAdaptiveOutputSpaceExploration"]
```
yes and yes.

quaquel commented on ema_workbench/em_framework/outputspace_exploration.py:
```diff
@@ -0,0 +1,409 @@
+    Parameters
+    ----------
+    problem : a platypus Problem instance
+    grid_spec : list of tuples
```
Technically, you could get away with just the epsilon values. For cross seed comparisons, you would need the a value for the origin of the grid structure, and the epsilon for the size of the grid. I had left it at three for now, but we can change this to a minimum (or origin) and epsilon value.

quaquel replied:
By the way, which examples do you mean with tuples of 2 values?

EwoutH replied:
A few places:

- https://github.com/quaquel/EMAworkbench/pull/170/files#diff-17ddc728ebc309d5ce247a5cb98dcfbb88047b15dfb9bcab0877770da2c3656cR120
- https://github.com/quaquel/EMAworkbench/pull/170/files#diff-14e09aadf560c254302311b901de243dbe999d7fca44a1b4321b0dc0505e05b3R133-R134
- https://github.com/quaquel/EMAworkbench/pull/170/files#diff-427cb03187f4e1f151fcb26edaf2d2752df9faef498122e6b11d10d18c3d78c7R151

EwoutH replied:
I read `0, 1, 0.1` as `0.1, 0.1`, whoops...

Maybe format as
```Python
grid_spec = [
    (0, 1, 0.1),
    (0, 1, 0.1),
]
```

quaquel replied:
yes (see my black comment in #178 )

quaquel commented on ema_workbench/em_framework/outputspace_exploration.py:
```diff
@@ -0,0 +1,409 @@
+    grid_spec : list of tuples
+                with min, max, and epsilon for
+                each outcome of interest
+    population_size : int, optional
```
100

quaquel commented on ema_workbench/em_framework/optimization.py:
```diff
@@ -776,50 +778,49 @@ def mutate_categorical(self, child, i, type):  # @ReservedAssignment
     _mutate = {Real: mutate_real, Integer: mutate_integer, Subset: mutate_categorical}
 
 
-class CombinedMutator(CombinedVariator):
```
yes that can go.

quaquel commented on ema_workbench/em_framework/outputspace_exploration.py:
```diff
@@ -0,0 +1,409 @@
+from Chérel et al, we have taken the idea of the HitBox. Basically, this is an epsilon archive where one
+keeps track of how many solutions have fallen into each grid cell. Next, tournament selection based on novelty is
+used as the selective pressure. Novelty is defined as 1/nr. of solutions in same grid cell. This is then
+combined with auto-adaptive population sizing as used in e-NSGAII. This replaces the use of adaptive Cauchy mutation
```
yes, and link is being added

quaquel commented:
I think this is now ready to be merged

EwoutH commented:
Awesome, I agree!

Can I merge #178? Then I will throw in some CI, rebase and check if #170 passes, and then merge it.

quaquel commented:
go ahead

EwoutH commented:
@quaquel It's in, congratulations on an amazing new feature!
